server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodServer.java

```diff
@@ -299,12 +301,30 @@ AdvancedCache getCacheInstance(String cacheName, EmbeddedCacheManager cacheManag
          Cache<byte[], byte[]> tmpCache = SecurityActions.getCache(cacheManager, validCacheName);
          Configuration cacheConfiguration = SecurityActions.getCacheConfiguration(tmpCache.getAdvancedCache());
          boolean compatibility = cacheConfiguration.compatibility().enabled();
+         Marshaller marshaller = cacheConfiguration.compatibility().marshaller();
          boolean indexing = cacheConfiguration.indexing().index().isEnabled();
+         ComponentRegistry cacheComponentRegistry = SecurityActions.getCacheComponentRegistry(tmpCache.getAdvancedCache());
+         if (marshaller != null) {
+            cacheComponentRegistry.wireDependencies(marshaller);
+         }
 
          // Use flag when compatibility is enabled, otherwise it's unnecessary
-         if (compatibility || indexing)
+         if (compatibility || indexing) {
             cache = tmpCache.getAdvancedCache().withFlags(Flag.OPERATION_HOTROD);
```

Comment:
Once this is complete, you should in theory be able to get rid of `OPERATION_` flags.